src/core/response_cache.cc

```diff
@@ -195,53 +167,135 @@ RequestResponseCache::Lookup(
   // Populate passed-in "ptr" from cache entry
   auto entry = iter->second;
   // Build InferenceResponse from CacheEntry
-  RETURN_IF_ERROR(BuildInferenceResponse(entry, ptr));
+  auto status = BuildInferenceResponse(entry, ptr);
+  if (!status.IsOk()) {
+    LOG_ERROR << "Failed to build inference response: " << status.Message();
+    return status;
+  }
 
   // Update this key to front of LRU list
   UpdateLRU(iter);
-
   LOG_VERBOSE(1) << "Using cached response for key [" + std::to_string(key) +
                         "].";
-
-  // Capture end lookup latency on hit and update total latency
-  request->CaptureCacheLookupEndNs();
-  total_lookup_latency_ns_ +=
-      (request->CacheLookupEndNs() - request->CacheLookupStartNs());
   return Status::Success;
 }
 
 Status
 RequestResponseCache::Insert(
-    const uint64_t key, const InferenceResponse& response)
+    const uint64_t key, const InferenceResponse& response,
+    InferenceRequest* request)
 {
   // Lock on cache insertion
   std::lock_guard<std::recursive_mutex> lk(cache_mtx_);
 
+  if (request == nullptr) {
+    return Status(
+        Status::Code::INTERNAL, "Cache Insert passed a nullptr request");
+  }
+
+  // Capture start latency now and end latency when timer goes out of scope
+  ScopedTimer timer(
+      *request, total_insertion_latency_ns_, ScopedTimerType::INSERTION);
+
   // Exit early if key already exists in cache
   auto iter = cache_.find(key);
   if (iter != cache_.end()) {
     return Status(
-        Status::Code::INTERNAL,
+        Status::Code::ALREADY_EXISTS,
         "key [" + std::to_string(key) + "] already exists in cache");
   }
 
   // Construct cache entry from response
   auto entry = CacheEntry();
-  RETURN_IF_ERROR(BuildCacheEntry(response, &entry));
```

Comment:
Same here. I think we can just leave it as :
RETURN_IF_ERROR(BuildCacheEntry(response, &entry));